File: app/src/main/java/jp/hazuki/yuzubrowser/provider/DownloadFileProvider.java
/*
 * Copyright (C) 2017-2018 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.provider;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.OpenableColumns;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

import jp.hazuki.utility.utils.ArrayUtils;
import jp.hazuki.yuzubrowser.BuildConfig;
import jp.hazuki.yuzubrowser.legacy.utils.FileUtils;

public class DownloadFileProvider extends ContentProvider {
    private static final String[] COLUMNS = {
            OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE};

    public static final String PATH = "file";

    private static final String SCHEME = "content";

    private static final String AUTHORITY = BuildConfig.APPLICATION_ID + ".downloadFileProvider";

    private String dataDir;

    @Override
    public boolean onCreate() {
        dataDir = getContext().getApplicationInfo().dataDir;
        return true;
    }

    @Nullable
    @Override
    public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {
        return ParcelFileDescriptor.open(getFileForUri(uri), modeToMode(mode));
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        File file = getFileForUri(uri);

        if (projection == null) {
            projection = COLUMNS;
        }

        String[] cols = new String[projection.length];
        Object[] values = new Object[projection.length];
        int i = 0;
        for (String col : projection) {
            if (OpenableColumns.DISPLAY_NAME.equals(col)) {
                cols[i] = OpenableColumns.DISPLAY_NAME;
                values[i++] = file.getName();
            } else if (OpenableColumns.SIZE.equals(col)) {
                cols[i] = OpenableColumns.SIZE;
                values[i++] = file.length();
            }
        }

        cols = ArrayUtils.copyOf(cols, i);
        values = ArrayUtils.copyOf(values, i);

        final MatrixCursor cursor = new MatrixCursor(cols, 1);
        cursor.addRow(values);
        return cursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return FileUtils.getMineType(getFileForUri(uri));
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        File file = getFileForUri(uri);
        return file.delete() ? 1 : 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @NonNull
    public static Uri getUriFromPath(String filePath) {
        String path = PATH + '/' + Uri.encode(filePath, "/");
        return new Uri.Builder().scheme(SCHEME).authority(AUTHORITY).encodedPath(path).build();
    }

    @NonNull
    public static Uri getUriForFile(File file) {
        String path;
        try {
            path = file.getCanonicalPath();
        } catch (IOException e) {
            throw new IllegalArgumentException("Failed to resolve canonical path for " + file);
        }
        return getUriFromPath(path);
    }

    private File getFileForUri(Uri uri) {
        String path = uri.getEncodedPath();

        final int splitIndex = path.indexOf('/', 1);
        final String tag = path.substring(1, splitIndex);
        path = Uri.decode(path.substring(splitIndex + 1));

        if (!PATH.equals(tag)) {
            throw new IllegalArgumentException("No files supported by provider at " + uri);
        }

        File file = new File(path);

        try {
            file = file.getCanonicalFile();
        } catch (IOException e) {
            throw new IllegalArgumentException("Failed to resolve canonical path for " + file);
        }

        if (file.getAbsolutePath().startsWith(dataDir)) {
            throw new IllegalArgumentException("No files supported by provider at " + uri);
        }
        return file;
    }

    /**
     * Copied from ContentResolver.java
     */
    private static int modeToMode(String mode) {
        int modeBits;
        if ("r".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_ONLY;
        } else if ("w".equals(mode) || "wt".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_WRITE_ONLY
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_TRUNCATE;
        } else if ("wa".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_WRITE_ONLY
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_APPEND;
        } else if ("rw".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_WRITE
                    | ParcelFileDescriptor.MODE_CREATE;
        } else if ("rwt".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_WRITE
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_TRUNCATE;
        } else {
            throw new IllegalArgumentException("Invalid mode: " + mode);
        }
        return modeBits;
    }
}

File: app/src/main/java/jp/hazuki/yuzubrowser/provider/DownloadFileProvider.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.provider

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.os.Environment
import android.os.ParcelFileDescriptor
import android.provider.OpenableColumns
import jp.hazuki.yuzubrowser.BuildConfig
import jp.hazuki.yuzubrowser.core.utility.utils.ArrayUtils
import jp.hazuki.yuzubrowser.core.utility.utils.FileUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

class DownloadFileProvider : ContentProvider() {

    private var dataDir: String? = null

    override fun onCreate(): Boolean {
        dataDir = context!!.applicationInfo.dataDir
        return true
    }

    @Throws(FileNotFoundException::class)
    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor? {
        val file = getFileForUri(uri)
        if (!file.checkPath()) return super.openFile(uri, mode)
        return ParcelFileDescriptor.open(file, modeToMode(mode))
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        var colRequest = projection
        val file = getFileForUri(uri)
        if (!file.checkPath()) return null

        if (colRequest == null) {
            colRequest = COLUMNS
        }

        var cols = arrayOfNulls<String>(colRequest.size)
        var values = arrayOfNulls<Any>(colRequest.size)
        var i = 0
        for (col in colRequest) {
            if (OpenableColumns.DISPLAY_NAME == col) {
                cols[i] = OpenableColumns.DISPLAY_NAME
                values[i++] = file.name
            } else if (OpenableColumns.SIZE == col) {
                cols[i] = OpenableColumns.SIZE
                values[i++] = file.length()
            }
        }

        cols = ArrayUtils.copyOf(cols, i)
        values = ArrayUtils.copyOf(values, i)

        val cursor = MatrixCursor(cols, 1)
        cursor.addRow(values)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        val file = getFileForUri(uri)
        if (!file.checkPath()) return null
        return FileUtils.getMineType(file)
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        throw UnsupportedOperationException()
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val file = getFileForUri(uri)
        return if (file.delete()) 1 else 0
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    private fun getFileForUri(uri: Uri): File {
        var path = uri.encodedPath

        val splitIndex = path!!.indexOf('/', 1)
        val tag = path.substring(1, splitIndex)
        path = Uri.decode(path.substring(splitIndex + 1))

        if (PATH != tag) {
            throw IllegalArgumentException("No files supported by provider at $uri")
        }

        var file = File(path!!)

        try {
            file = file.canonicalFile
        } catch (e: IOException) {
            throw IllegalArgumentException("Failed to resolve canonical path for $file")
        }

        if (file.absolutePath.startsWith(dataDir!!)) {
            throw IllegalArgumentException("No files supported by provider at $uri")
        }
        return file
    }

    private fun File.checkPath(): Boolean {
        return isFile &&
                path.startsWith("/storage/") ||
                path.startsWith(Environment.getExternalStorageDirectory().absolutePath) ||
                path.startsWith("/mnt/")
    }

    companion object {
        private val COLUMNS = arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE)

        const val PATH = "file"

        private const val SCHEME = "content"

        private const val AUTHORITY = BuildConfig.APPLICATION_ID + ".downloadFileProvider"

        fun getUriFromPath(filePath: String): Uri {
            val path = PATH + '/'.toString() + Uri.encode(filePath, "/")
            return Uri.Builder().scheme(SCHEME).authority(AUTHORITY).encodedPath(path).build()
        }

        fun getUriForFile(file: File): Uri {
            val path: String
            try {
                path = file.canonicalPath
            } catch (e: IOException) {
                throw IllegalArgumentException("Failed to resolve canonical path for $file")
            }

            return getUriFromPath(path)
        }

        /**
         * Copied from ContentResolver.java
         */
        private fun modeToMode(mode: String): Int {
            val modeBits: Int
            if ("r" == mode) {
                modeBits = ParcelFileDescriptor.MODE_READ_ONLY
            } else if ("w" == mode || "wt" == mode) {
                modeBits = (ParcelFileDescriptor.MODE_WRITE_ONLY
                        or ParcelFileDescriptor.MODE_CREATE
                        or ParcelFileDescriptor.MODE_TRUNCATE)
            } else if ("wa" == mode) {
                modeBits = (ParcelFileDescriptor.MODE_WRITE_ONLY
                        or ParcelFileDescriptor.MODE_CREATE
                        or ParcelFileDescriptor.MODE_APPEND)
            } else if ("rw" == mode) {
                modeBits = ParcelFileDescriptor.MODE_READ_WRITE or ParcelFileDescriptor.MODE_CREATE
            } else if ("rwt" == mode) {
                modeBits = (ParcelFileDescriptor.MODE_READ_WRITE
                        or ParcelFileDescriptor.MODE_CREATE
                        or ParcelFileDescriptor.MODE_TRUNCATE)
            } else {
                throw IllegalArgumentException("Invalid mode: $mode")
            }
            return modeBits
        }
    }
}
